third_party/move/tools/move-package-resolver/src/resolver.rs

```diff
@@ -0,0 +1,438 @@
+
+/// Checks for cyclic dependencies in the given resolution graph.
+fn check_for_cyclic_dependencies(graph: &ResolutionGraph) -> Result<()> {
+    let format_scc = |scc: &[NodeIndex]| {
+        scc.iter()
+            .map(|node| {
+                format!(
+                    "{} @ {}",
+                    graph[*node].identity.name, graph[*node].identity.location
+                )
+            })
+            .collect::<Vec<_>>()
+            .join("\n")
+    };
+
+    let sccs = kosaraju_scc(graph)
+        .into_iter()
+        .filter(|scc| scc.len() > 1)
+        .collect::<Vec<_>>();
+
+    if !sccs.is_empty() {
+        let sccs = sccs.iter().map(|scc| format_scc(scc)).collect::<Vec<_>>();
+        bail!("Cyclic dependencies found:\n{}", sccs.join("\n\n"));
+    }
+
+    Ok(())
+}
+
+/// Checks if any node has an edge to itself -- this is a special form of cyclic dependency.
+fn check_for_self_dependencies(graph: &ResolutionGraph) -> Result<()> {
+    let mut result = Vec::new();
+
+    for edge in graph.edge_references() {
+        if edge.source() == edge.target() {
+            result.push(edge.source());
+        }
+    }
+
+    result.sort_unstable();
+    result.dedup();
+
+    if !result.is_empty() {
+        bail!(
+            "Found packages with self-dependencies:\n{}",
+            result
+                .iter()
+                .map(|idx| format!(
+                    "{} @ {}",
+                    graph[*idx].identity.name, graph[*idx].identity.location
+                ))
+                .collect::<Vec<_>>()
+                .join("\n")
+        );
+    }
+
+    Ok(())
+}
+
+/// Checks if two different packages have the same name -- for now we forbid this, but
+/// plan to relax it in the future.
+fn check_for_name_conflicts(graph: &ResolutionGraph) -> Result<()> {
+    let mut name_location_map = BTreeMap::new();
+
+    for node in graph.node_indices() {
+        let identity = &graph[node].identity;
+
+        let locations = name_location_map
+            .entry(identity.name.as_str())
+            .or_insert_with(Vec::new);
+        locations.push(&identity.location);
+    }
+
+    let conflicts = name_location_map
+        .into_iter()
+        .filter(|(_name, locations)| locations.len() > 1)
+        .map(|(name, locations)| {
+            format!(
+                "Package name conflict: {}\n{}",
+                name,
+                locations
+                    .iter()
+                    .map(|l| format!("  {}", l))
+                    .collect::<Vec<_>>()
+                    .join("\n")
+            )
+        })
+        .collect::<Vec<_>>()
+        .join("\n\n");
+
+    if !conflicts.is_empty() {
+        bail!("{}", conflicts);
+    }
+
+    Ok(())
+}
+
+/// Resolves all transitive dependencies for the given root package.
+/// The results are returned as a [`ResolutionGraph`].
+///
+/// During resolution, remote dependencies are fetched and cached.
+pub async fn resolve(
```

Comment:
Given this is a public function, would be good to document how `dev_mode` affects the behavior.